ecs-agent/acs/session/session.go

```diff
@@ -11,13 +11,348 @@
+		NewRefreshCredentialsResponder(s.credentialsManager, s.credentialsMetadataSetter, s.metricsFactory,
+			responseSender),
+		NewAttachTaskENIResponder(s.eniHandler, responseSender),
+		NewAttachInstanceENIResponder(s.eniHandler, responseSender),
+		NewHeartbeatResponder(s.doctor, responseSender),
+		NewTaskManifestResponder(s.taskComparer, s.sequenceNumberAccessor, s.manifestMessageIDAccessor,
+			s.metricsFactory, responseSender),
+		NewTaskStopVerificationACKResponder(s.taskStopper, s.manifestMessageIDAccessor, s.metricsFactory),
+	}
+	for _, r := range responders {
+		client.AddRequestHandler(r.HandlerFunc())
+	}
+
+	if s.dockerVersion == "containerd" && s.resourceHandler != nil {
+		client.AddRequestHandler(NewAttachResourceResponder(s.resourceHandler, s.metricsFactory,
+			responseSender).HandlerFunc())
+	}
+
+	if s.dockerVersion != "containerd" && s.addUpdateRequestHandlers != nil {
+		s.addUpdateRequestHandlers(client)
+	}
+
+	// Start a heartbeat timer for closing the connection.
+	heartbeatTimer := newHeartbeatTimer(client, s.heartbeatTimeout, s.heartbeatJitter)
+	// Any message from the server resets the heartbeat timer.
+	client.SetAnyRequestHandler(anyMessageHandler(heartbeatTimer, client))
+	defer heartbeatTimer.Stop()
+
+	backoffResetTimer := time.AfterFunc(
+		retry.AddJitter(s.heartbeatTimeout, s.heartbeatJitter), func() {
+			// If we do not have an error connecting and remain connected for at
+			// least 1 or so minutes, reset the backoff. This prevents disconnect
+			// errors that only happen infrequently from damaging the reconnect
+			// delay as significantly.
+			s.backoff.Reset()
+		})
+	defer backoffResetTimer.Stop()
+
+	return client.Serve(ctx)
+}
+
+func (s *session) reconnectDelay(acsError error) (time.Duration, bool) {
+	if isInactiveInstanceError(acsError) {
+		logger.Info("Container instance is deregistered")
+		s.inactiveInstanceCB()
+		return s.inactiveInstanceReconnectDelay, true
+	}
+	if shouldReconnectWithoutBackoff(acsError) {
+		// ACS has closed the connection for valid reasons. Example: periodic disconnect.
+		// No need to wait/backoff to reconnect.
+		logger.Info("ACS WebSocket connection closed for a valid reason")
+		s.backoff.Reset()
+		return 0, false
+
+	}
+	// Disconnected unexpectedly from ACS, compute backoff duration to reconnect.
+	return s.backoff.Duration(), true
+}
+
+// acsURL returns the websocket url for ACS given the endpoint.
+func (s *session) acsURL(endpoint string) string {
+	wsURL := endpoint
+	if endpoint[len(endpoint)-1] != '/' {
+		wsURL += "/"
+	}
+	wsURL += "ws"
+	query := url.Values{}
+	query.Set("clusterArn", s.cluster)
+	query.Set("containerInstanceArn", s.containerInstanceARN)
+	query.Set("agentHash", s.agentHash)
+	query.Set("agentVersion", s.agentVersion)
+	query.Set("seqNum", "1")
+	query.Set("protocolVersion", strconv.Itoa(acsProtocolVersion))
+	if s.dockerVersion != "" {
+		dockerVersionValue := s.dockerVersion
+		if s.dockerVersion != "containerd" {
```

Comment:
Huh so if I understand correctly Fargate now passes "containerd" to this? I don't have a solid idea about this now and might be an AI for future, but I feel like in case we want to get adoption metrics for containerd version this implementation can be blocking. In TACS I had an extra parameter as container runtime and check that.

Reply:
that is correct, I think the future goal should be to pass in a `daemon` attribute (containerd or docker) and a corresponding `runtimeVersion`. Getting containerd version could be a follow up task but it is unclear if this data is already exposed to ACS. Putting the API link here for brevity ([Ref](https://github.com/containerd/containerd/blob/main/client.go#L754-L770))

Reply:
Discussed offline - changes required for getting adoption metrics for containerd version in the future if we do want to are small and will be out of scope of this PR. They will be included as part of any PR in the future to include containerd version as part of the websocket URL for ACS and TACS connections.